src/pstore/pstore.c

```diff
@@ -0,0 +1,399 @@
+        iovec[n_iovec++] = IOVEC_MAKE_STRING(field);
+
+        field_size = strlen("FILE=") + pe->content_size;
+        field=malloc(field_size);
+        if (!field)
+                return log_oom();
+        memcpy(field, "FILE=", strlen("FILE="));
+        memcpy(field + strlen("FILE="), pe->content, pe->content_size);
+        iovec[n_iovec++] = IOVEC_MAKE(field, field_size);
+
+        r = sd_journal_sendv(iovec, n_iovec);
+        if (r < 0)
+                return log_error_errno(r, "Failed to log pstore entry: %m");
+        free(field);
+
+        if (arg_storage == PSTORE_STORAGE_EXTERNAL) {
+                /* Move file from pstore to external storage */
+                r = mkdir_parents(ofd_path, 0755);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to create directoy %s: %m", ofd_path);
+                r = copy_file_atomic(ifd_path, ofd_path, 0644, 0, 0, COPY_REPLACE);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to copy_file_atomic: %s to %s", ifd_path, ofd_path);
+        }
+
+        /* If file copied properly, remove it from pstore */
+        if (arg_unlink)
+                (void) unlink(ifd_path);
+
+        return 0;
+}
+
+static int write_dmesg(const char *dmesg, size_t size, const char *id) {
+        _cleanup_free_ char *ofd_path;
+        _cleanup_free_ char *tmp_path = NULL;
+        _cleanup_close_ int ofd = -1;
+        ssize_t wr;
+        int r;
+
+        if (isempty(dmesg) || (size == 0))
+                return 0;
+
+        log_info("Record ID %s", id);
+
+        ofd_path = path_join(arg_archivedir, id, "dmesg.txt");
+        if (!ofd_path)
+                return log_oom();
+
+        ofd = open_tmpfile_linkable(ofd_path, O_CLOEXEC|O_CREAT|O_TRUNC|O_WRONLY, &tmp_path);
+        if (ofd < 0)
+                return log_error_errno(ofd, "Failed to open temporary file %s: %m", ofd_path);
+        wr = write(ofd, dmesg, size);
+        if (wr < 0)
+                return log_error_errno(errno, "Failed to store dmesg to %s: %m", ofd_path);
+        if (wr != (ssize_t)size)
+                return log_error("Failed to store dmesg to %s. %zu bytes are lost.", ofd_path, size - wr);
+        r = link_tmpfile(ofd, tmp_path, ofd_path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write temporary file %s: %m", ofd_path);
+
+        return 0;
+}
+
+static void process_dmesg_files(PStoreList *list) {
+        /* Move files, reconstruct dmesg.txt */
+        PStoreEntry *pe;
+        char *dmesg = NULL, *dmesg2;
+        size_t dmesg_size = 0;
+        char id1[ENTRYSZ], id2[ENTRYSZ], *id;
+        char *current_id;
+        char *p;
+
+        id1[0] = id2[0] = '\0';
+        id = id1;
+
+        /* Handle each dmesg file: files processed in reverse
+         * order so as to properly reconstruct original dmesg */
+        for (int n = list->n_entries; n > 0; n--) {
```

Comment:
type mismatch: list->n_entries is `size_t` (and rightfully so, since we count memory objects), hence this iterator should be `size_t n` too here, not `int n`

Reply:
done